light/lightchain.go

```diff
@@ -459,15 +554,48 @@ func (lc *LightChain) GetHeaderByHash(hash common.Hash) *types.Header {
 	return lc.hc.GetHeaderByHash(hash)
 }
 
+// GetHeaderByNumberWithContext retrieves a block header from the database or network
+// by number, caching it (associated with its hash) if found.
+func (lc *LightChain) GetHeaderByNumberWithContext(ctx context.Context, number uint64) (*types.Header, error) {
+	if header := lc.hc.GetHeaderByNumber(number); header != nil {
+		return header, nil
+	}
+	return GetHeaderByNumber(ctx, lc.odr, number)
+}
+
+// GetHeaderByNumber retrieves a block header from the database by number,
+// caching it (associated with its hash) if found.
+func (lc *LightChain) GetHeaderByNumber(number uint64) *types.Header {
+	header, err := lc.GetHeaderByNumberWithContext(DefaultContext, number)
+	if err != nil {
+		return nil
+	}
+	return header
+}
+
 // HasHeader checks if a block header is present in the database or not, caching
 // it if present.
 func (lc *LightChain) HasHeader(hash common.Hash, number uint64) bool {
-	return lc.hc.HasHeader(hash, number)
+	header, _ := lc.GetHeaderWithContext(DefaultContext, hash, number)
+	return header != nil
+}
+
+// GetCanonicalHashWithContext returns the canonical hash for a given block number
+func (lc *LightChain) GetCanonicalHashWithContext(ctx context.Context, number uint64) (common.Hash, error) {
+	hash := lc.hc.GetCanonicalHash(number)
+	if hash != (common.Hash{}) {
+		return hash, nil
+	}
+	return GetCanonicalHash(ctx, lc.odr, number)
 }
 
 // GetCanonicalHash returns the canonical hash for a given block number
-func (bc *LightChain) GetCanonicalHash(number uint64) common.Hash {
-	return bc.hc.GetCanonicalHash(number)
+func (lc *LightChain) GetCanonicalHash(number uint64) common.Hash {
+	hash, err := lc.GetCanonicalHashWithContext(DefaultContext, number)
```

Comment:
This does not look right. I mean it is very good to get rid of the "NoOdr" hack because it was misleading, it does not disable the ODR (it was mostly used in tests where the ODR is also explicitly disabled in the test backend). I also like the consistent naming of functions. But now you implemented the non-ODR versions of the LightChain functions by calling the ODR version with DefaultContext (which still DOES use ODR and will block forever if the object is not accessible). Is there something I'm missing here?

Reply:
> But now you implemented the non-ODR versions of the LightChain functions by calling the ODR version with DefaultContext (which still DOES use ODR and will block forever if the object is not accessible)

Yes. It's the core change. In this approach, the `LightChain` can have a unified interface with `core.Blockchain`.
But it's true that if we promote this type of functions(call the default context), the `timeout` mechanism or ODR is useless. Let me think about it.